File: backend/schemas/task.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class TaskDraftRequest(BaseModel):
    title: str
    description: str = ""
    grading_criteria: str = ""
    class_id: uuid.UUID


class TaskUpdateRequest(BaseModel):
    title: str | None = None
    description: str | None = None
    grading_criteria: str | None = None
    status: str | None = None


class TaskResponse(BaseModel):
    id: uuid.UUID
    title: str
    description: str
    grading_criteria: str
    status: str
    class_id: uuid.UUID
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime | None
    class_name: str
    created_by_name: str


class TaskListResponse(BaseModel):
    items: list[TaskResponse]


class TaskSubmissionItem(BaseModel):
    student_id: uuid.UUID
    username: str
    version: int
    submission_count: int
    status: str
    score: float | None
    submitted_at: datetime


class TaskStatsResponse(BaseModel):
    task_id: uuid.UUID
    total_students: int
    submitted_count: int
    submission_rate: float
    submissions: list[TaskSubmissionItem]
    not_submitted: list[str]


class GenerateCriteriaRequest(BaseModel):
    title: str
    description: str


class GenerateCriteriaResponse(BaseModel):
    criteria: str


class BatchPublishRequest(BaseModel):
    title: str
    description: str
    grading_criteria: str
    class_ids: list[uuid.UUID]
    status: Literal["published"]


class BatchPublishItem(BaseModel):
    id: uuid.UUID
    class_id: uuid.UUID
    class_name: str


class BatchPublishResponse(BaseModel):
    created: list[BatchPublishItem]

File: my-app/src/types/task.ts
export interface LearningResource {
  url: string;
  title: string;
  content: string;
}

export interface Task {
  id: string;
  title: string;
  description: string;
  grading_criteria: string;
  learning_resources: LearningResource[] | null;
  status: "draft" | "published";
  class_id: string;
  created_by: string;
  created_at: string;
  updated_at: string | null;
  class_name: string;
  created_by_name: string;
}

export interface TaskListResponse {
  items: Task[];
}

export interface TaskDraftRequest {
  title: string;
  description?: string;
  grading_criteria?: string;
  learning_resources?: LearningResource[] | null;
  class_id: string;
}

export interface TaskUpdateRequest {
  title?: string;
  description?: string;
  grading_criteria?: string;
  learning_resources?: LearningResource[] | null;
  status?: "published";
}

export interface TaskSubmissionItem {
  student_id: string;
  username: string;
  version: number;
  submission_count: number;
  status: string;
  score: number | null;
  submitted_at: string;
}

export interface TaskStatsResponse {
  task_id: string;
  total_students: number;
  submitted_count: number;
  submission_rate: number;
  submissions: TaskSubmissionItem[];
  not_submitted: string[];
}

export interface GenerateCriteriaRequest {
  title: string;
  description: string;
}

export interface GenerateCriteriaResponse {
  criteria: string;
}

export interface BatchPublishRequest {
  title: string;
  description: string;
  grading_criteria: string;
  learning_resources?: LearningResource[] | null;
  class_ids: string[];
  status: "published";
}

export interface BatchPublishResponse {
  created: { id: string; class_id: string; class_name: string }[];
}
